From 8d281df8fc5e7008f2262d989d6fd27e001971d9 Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Tue, 30 Apr 2019 16:21:11 +0300
Subject: [PATCH] [SW-0] pll: introduce hl_pll_get_freq()

The hl_pll_get_freq() function calculates and returns the output
frequency of the requested PLL divider output.
The calculation takes into an account the DIV_SEL configuration
(both the reference clock and other options) as well as the PLL
coefficients.

Change-Id: I21d5c9c2b8f88d1c6d8ff726977fd6ba4e28a3c3
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/common/pll.c b/common/pll.c
index 6e26e4c..8025a17 100644
--- a/common/pll.c
+++ b/common/pll.c
@@ -162,6 +162,34 @@
 	return ret;
 }
 
+u32 hl_pll_get_freq(u64 pll_base, enum hl_pll_div_out div)
+{
+	u32 nr, nf, od, div_sel, factor, offset;
+	u64 out_freq = SYS_REF_CLOCK;
+
+	offset = div * sizeof(u32);
+	div_sel = regmap_readl(pll_base + HL_PLL_DIV_SEL + offset);
+
+	if (div_sel == HL_PLL_PLL_CLK || div_sel == HL_PLL_DIV_PLL) {
+		nr = regmap_readl(pll_base + HL_PLL_NR) + 1;
+		nf = regmap_readl(pll_base + HL_PLL_NF) + 1;
+		od = regmap_readl(pll_base + HL_PLL_OD) + 1;
+		out_freq = (SYS_REF_CLOCK * nf) / (nr * od);
+	}
+
+	if (div_sel == HL_PLL_DIV_REF || div_sel == HL_PLL_DIV_PLL) {
+		factor = regmap_readl(pll_base + HL_PLL_DIV_FACTOR + offset);
+		out_freq /= factor + 1;
+	}
+
+	/*
+	 * The clock relaxation is not accounted here.
+	 * We might need to deal with it in the future.
+	 */
+
+	return out_freq;
+}
+
 /**
  * hl_pll_init - Initialize PLL registers
  */
diff --git a/common/pll.h b/common/pll.h
index 56c6641..4cce062 100644
--- a/common/pll.h
+++ b/common/pll.h
@@ -61,6 +61,8 @@
 int hl_pll_get_cfg(struct hl_pll_cfg **hl_pll_cfg);
 u64 hl_pll_get_cpu_cycles_per_usec(void);
 u64 hl_pll_get_cpu_pll_base_off(void);
+u32 hl_pll_get_freq(u64 pll_base, enum hl_pll_div_out div);
+
 int hl_pll_partial_init_get(void);
 void hl_pll_partial_init_set(void);
 
